modules/gateway/peers.go

```diff
@@ -1,42 +1,129 @@
 package gateway
 
 import (
-	"github.com/NebulousLabs/Sia/network"
+	"errors"
+	"math/rand"
+	"sync"
+
+	"github.com/NebulousLabs/Sia/modules"
 )
 
-// SharePeers returns up to 10 randomly selected peers.
-func (g *Gateway) SharePeers() (peers []network.Address, err error) {
-	g.mu.RLock()
-	defer g.mu.RUnlock()
+const (
+	maxSharedPeers = 10
+	maxAddrLength  = 100
+)
+
+func (g *Gateway) addPeer(peer modules.NetAddress) error {
+	if _, exists := g.peers[peer]; exists {
+		return errors.New("peer already added")
+	}
+	g.peers[peer] = 0
+	return nil
+}
+
+func (g *Gateway) removePeer(peer modules.NetAddress) error {
+	if _, exists := g.peers[peer]; !exists {
+		return errors.New("no record of that peer")
+	}
+	delete(g.peers, peer)
+	return nil
+}
+
+func (g *Gateway) randomPeer() (modules.NetAddress, error) {
+	r := rand.Intn(len(g.peers))
 	for peer := range g.peers {
-		if len(peers) > 10 {
-			return
+		if r == 0 {
+			return peer, nil
 		}
-		peers = append(peers, peer)
+		r--
 	}
-	return
+	return "", ErrNoPeers
 }
 
-// AddPeer is an RPC that requests that the Gateway add a peer to its peer
-// list. The supplied peer is assumed to be the peer making the RPC.
-func (g *Gateway) AddMe(peer network.Address) error {
-	if !network.Ping(peer) {
-		return ErrUnreachable
+func (g *Gateway) addStrike(peer modules.NetAddress) {
+	if _, exists := g.peers[peer]; !exists {
+		return
+	}
+	g.peers[peer]++
+	if g.peers[peer] > maxStrikes {
+		g.removePeer(peer)
 	}
-	g.AddPeer(peer)
-	return nil
 }
 
 // AddPeer adds a peer to the Gateway's peer list.
-func (g *Gateway) AddPeer(peer network.Address) error {
+func (g *Gateway) AddPeer(peer modules.NetAddress) error {
 	g.mu.Lock()
 	defer g.mu.Unlock()
 	return g.addPeer(peer)
 }
 
 // RemovePeer removes a peer from the Gateway's peer list.
-func (g *Gateway) RemovePeer(peer network.Address) error {
+func (g *Gateway) RemovePeer(peer modules.NetAddress) error {
 	g.mu.Lock()
 	defer g.mu.Unlock()
 	return g.removePeer(peer)
 }
+
+// addMe is an RPC that requests that the Gateway add a peer to its peer
+// list. The supplied peer is assumed to be the peer making the RPC.
+func (g *Gateway) addMe(conn modules.NetConn) error {
+	var peer modules.NetAddress
+	err := conn.ReadObject(&peer, maxAddrLength)
+	if err != nil {
+		return err
+	}
+	if !g.Ping(peer) {
+		return ErrUnreachable
+	}
+	g.AddPeer(peer)
+	return nil
+}
+
+// sharePeers is an RPC that returns up to 10 randomly selected peers.
+func (g *Gateway) sharePeers(conn modules.NetConn) error {
+	g.mu.RLock()
+	defer g.mu.RUnlock()
```

Comment:
RPCs are a special case, because they are, in fact, "exported" in the sense that they can be called externally (by another peer). So the locks are definitely needed. It definitely needs a docstring explaining why, though.